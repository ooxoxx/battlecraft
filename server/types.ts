import { z } from 'zod'

export enum Dungeon {
  AraKaraCityOfEchoes = '12660', // Ara-Kara, City of Echoes
  CityOfThreads = '12669', // City of Threads
  GrimBatol = '60670', // Grim Batol
  MistsOfTirnaScithe = '62290', // Mists of Tirna Scithe
  SiegeOfBoralus = '61822', // Siege of Boralus
  TheDawnbreaker = '12662', // The Dawnbreaker
  TheNecroticWake = '62286', // The Necrotic Wake
  TheStonevault = '12652', // The Stonevault
}

export enum CharacterClass {
  DeathKnight = 'DeathKnight',
  Druid = 'Druid',
  Hunter = 'Hunter',
  Mage = 'Mage',
  Monk = 'Monk',
  Paladin = 'Paladin',
  Priest = 'Priest',
  Rogue = 'Rogue',
  Shaman = 'Shaman',
  Warlock = 'Warlock',
  Warrior = 'Warrior',
  DemonHunter = 'DemonHunter',
  Evoker = 'Evoker',
}

export enum Specialization {
  Blood = 'Blood',
  Frost = 'Frost',
  Unholy = 'Unholy',
  Balance = 'Balance',
  Feral = 'Feral',
  Guardian = 'Guardian',
  Restoration = 'Restoration',
  BeastMastery = 'BeastMastery',
  Marksmanship = 'Marksmanship',
  Survival = 'Survival',
  Arcane = 'Arcane',
  Fire = 'Fire',
  Brewmaster = 'Brewmaster',
  Mistweaver = 'Mistweaver',
  Windwalker = 'Windwalker',
  Holy = 'Holy',
  Protection = 'Protection',
  Retribution = 'Retribution',
  Discipline = 'Discipline',
  Shadow = 'Shadow',
  Assassination = 'Assassination',
  Subtlety = 'Subtlety',
  Outlaw = 'Outlaw',
  Elemental = 'Elemental',
  Enhancement = 'Enhancement',
  Affliction = 'Affliction',
  Demonology = 'Demonology',
  Destruction = 'Destruction',
  Arms = 'Arms',
  Fury = 'Fury',
  Havoc = 'Havoc',
  Vengeance = 'Vengeance',
  Devastation = 'Devastation',
  Preservation = 'Preservation',
  Augmentation = 'Augmentation',
}

export interface CharacterStats {
  'strength'?: number
  'agility'?: number
  'intellect'?: number
  'stamina': number
  'leech': number
  'armor': number
  'mastery': number
  'haste': number
  'crit': number
  'versatility': number
  'item-level': number // note the quoted key for compatibility
  'avoidance': number
  'speed': number
  'block': number
  'parry': number
  'dodge': number
}

export const CharacterClassEnum = z.nativeEnum(CharacterClass)
export const SpecializationEnum = z.nativeEnum(Specialization)
export const DungeonEnum = z.nativeEnum(Dungeon)
