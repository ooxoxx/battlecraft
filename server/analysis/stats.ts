import type { CharacterClass, CharacterStats, Dungeon, Specialization } from '../types'
import { JSONPath } from 'jsonpath-plus'
import pLimit from 'p-limit'
import { classesAndSpecs, dungeons, toCompact } from '../../app/constants'
import { gql } from '../__generated__/gql'
import { useClientWithEnv } from '../apollo-client'
import { CharacterClassEnum, DungeonEnum, SpecializationEnum } from '../types'

const CHARACTER_RANKING = gql(/* GraphQL */`
  query characterRanking(
    $id: Int = 12660
    $className: String!
    $specName: String!
  ) {
    worldData {
      encounter(id: $id) {
        name
        characterRankings(
          page: 1
          className: $className
          specName: $specName
          metric: playerscore
          leaderboard: LogsOnly
        )
      }
    }
  }
`)

interface BySpecDungeonArgs {
  className: string
  specName: string
  dungeon: string
}

interface PlayerDetails {
  name: string
  class: string
  spec: string
  amount: number
  hardModeLevel: number
  duration: number
  startTime: number
  report: {
    code: string
    fightID: number
    startTime: number
  }
  server: {
    id: number
    name: string
  }
  bracketData: number
  faction: number
  affixes: number[]
  medal: string
  score: number
  leaderboard: number
  hidden?: boolean
}

export async function rankingsBySpecDungeon({ className, specName, dungeon }: BySpecDungeonArgs): Promise<PlayerDetails[]> {
  const client = await useClientWithEnv()
  const data = await client.query({ query: CHARACTER_RANKING, variables: {
    className,
    specName,
    dungeon,
  } })
  return data.data?.worldData?.encounter?.characterRankings?.rankings
}

const REPORT_SUMMARY = gql(/* GraphQL */`
  query ReportSummary($code: String!, $fight: Int!) {
    reportData {
      report(code: $code) {
        fights(fightIDs: [$fight]) {
          id
          encounterID
        }
        code
        table(fightIDs: [$fight], dataType: Summary)
      }
    }
}`)

interface StatsByPlayerReportsArgs {
  name: string
  server: string
  code: string
  fight: number
}

export interface TopStats {
  name: string
  rank: number
  class: CharacterClass
  spec: Specialization
  dungeon: Dungeon
  stats: CharacterStats
}

interface StatRange {
  min: number
  max: number
}

interface PlayerStats {
  'Speed': StatRange
  'Mastery': StatRange
  'Crit': StatRange
  'Item Level': StatRange
  'Stamina': StatRange
  'Strength'?: StatRange
  'Agility'?: StatRange
  'Intellect'?: StatRange
  'Avoidance': StatRange
  'Haste': StatRange
  'Versatility': StatRange
  'Leech': StatRange
  'Armor': StatRange
}

// Type for CombatantInfo (Combatant-specific details)
interface CombatantInfo {
  stats: PlayerStats
}

// Type for PlayerSummary (General player information)
interface PlayerSummary {
  name: string
  id: number
  guid: number
  type: CharacterClass // Enum for character class
  server: string
  icon: string
  specs: Specialization[] // Array of specializations (using the Specialization enum)
  minItemLevel: number
  maxItemLevel: number
  potionUse: number
  healthstoneUse: number
  combatantInfo: CombatantInfo | [] // Nested combatant info
}

export const statsByPlayerReport = defineCachedFunction< Promise<TopStats | undefined>, [StatsByPlayerReportsArgs]>(
  async ({ name, server, code, fight }) => {
    const client = await useClientWithEnv()
    const data = await client.query({ query: REPORT_SUMMARY, variables: { code, fight } })
    // console.log(data.data.reportData.report.code, data.data.reportData.report.fights[0].id)
    // console.log(client.cache.identify(data.data.reportData))

    // find the player in report
    const players = data.data?.reportData?.report?.table.data.playerDetails as { tanks: PlayerSummary[], healers: PlayerSummary[], dps: PlayerSummary[] }
    server = server.replace(' ', '')
    const player = [players.tanks, players.dps, players.healers].flat().find(p => p.name === name && p.server === server)

    if (player === undefined) {
      throw new Error('Player not found in report')
    }

    if (Array.isArray(player.combatantInfo)) {
      return
    }

    // Extract keys and min values using JSONPath
    const keys = JSONPath<string[]>({ path: '$.*~', json: player.combatantInfo.stats }) // Extract keys
    const minValues = JSONPath<number[]>({ path: '$.*.min', json: player.combatantInfo.stats }) // Extract min values

    // Combine keys and values into a single object
    const stats = keys?.reduce((acc, key, index) => {
      key = key.toLowerCase().replace(' ', '-') // handle 'item level'
      // @ts-expect-error: acc maybe of any type
      acc[key] = minValues[index]
      return acc
    }, {}) as CharacterStats

    const encounterID = data.data?.reportData?.report?.fights?.at(0)?.encounterID

    return {
      name,
      class: player.type,
      spec: player.specs[0],
      dungeon: DungeonEnum.parse(encounterID?.toString()),
      rank: -1,
      stats,
    }
  },
  {
    maxAge: 365 * 60 * 60 * 24,
    staleMaxAge: -1,
    name: 'statsByPlayerReport',
    getKey: ({ name, server, code, fight }) => `${name}-${server}-${code}-${fight}`,
  },
)

export const topStatsBySpecDungeon = defineCachedFunction<Promise<TopStats[]>, [BySpecDungeonArgs]>(
  async ({ className, specName, dungeon }) => {
    className = CharacterClassEnum.parse(toCompact(className))
    specName = SpecializationEnum.parse(toCompact(specName))
    dungeon = DungeonEnum.parse(dungeon)
    const details = await rankingsBySpecDungeon({ className, specName, dungeon })

    const top5Stats = details.slice(0, 7)

    // note that Promise.all preserve the order of its input array and output array
    return Promise.all(top5Stats.filter(ts => !ts.hidden).map(async (ts, i) => {
      const stats = await statsByPlayerReport({ name: ts?.name, server: ts.server.name, code: ts.report.code, fight: ts.report.fightID })
      if (stats === undefined) {
        return
      }
      return {
        ...stats,
        rank: i,
      }
    })).then(stats => stats.filter((s): s is TopStats => s !== undefined).slice(0, 5))
  },
  {
    name: 'topStatsBySpecDungeon',
    maxAge: 60 * 60,
    staleMaxAge: -1,
    getKey: async ({ className, specName, dungeon }) => {
      return `${className}-${specName}-${dungeon}`
    },
  },
)

export async function aquireTopStats() {
  // console.log('aquire entered')
  const limit = pLimit(10)
  await Promise.all(classesAndSpecs.map(
    ({ name, specs }) => specs.map(
      spec => dungeons.map(
        ({ id: dungID }) => limit(async () => {
          let counter = 0
          while (counter < 3) {
            try {
              // console.log(`aquiring topstats of class: ${name}, spec: ${spec.name}, dungeon: ${dungID}`)
              await topStatsBySpecDungeon({
                className: name,
                specName: spec.name,
                dungeon: dungID,
              })
            }
            catch (e) {
              console.warn(`Failed to fetch topstats ${name} ${spec} ${dungID} remaining ${3 - counter} retries`)
              console.warn(e)
              counter++
              continue
            }
            // console.log(`aquirie done topstats of class: ${name}, spec: ${spec.name}, dungeon: ${dungID}`)
            break
          }
        }),
      ),
    ),
  ).flat(3))
}
