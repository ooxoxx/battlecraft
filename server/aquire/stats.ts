import type { CharacterClass, Dungeon, Specialization } from '../types'
import { JSONPath } from 'jsonpath-plus'
import { gql } from '../__generated__/gql'
import { useClientWithEnv } from '../apollo-client'

const CHARACTER_RANKING = gql(/* GraphQL */`
  query characterRanking(
    $id: Int = 12660
    $className: String!
    $specName: String!
  ) {
    worldData {
      encounter(id: $id) {
        name
        characterRankings(
          page: 1
          className: $className
          specName: $specName
          metric: playerscore
          leaderboard: LogsOnly
        )
      }
    }
  }
`)

interface BySpecDungeonArgs {
  className: CharacterClass
  specName: Specialization
  dungeon: Dungeon
}

interface PlayerDetails {
  name: string
  class: string
  spec: string
  amount: number
  hardModeLevel: number
  duration: number
  startTime: number
  report: {
    code: string
    fightID: number
    startTime: number
  }
  server: {
    id: number
    name: string
  }
  bracketData: number
  faction: number
  affixes: number[]
  medal: string
  score: number
  leaderboard: number
}

export async function rankingsBySpecDungeon({ className, specName, dungeon }: BySpecDungeonArgs): Promise<PlayerDetails[]> {
  const client = await useClientWithEnv()
  const data = await client.query({ query: CHARACTER_RANKING, variables: {
    className,
    specName,
    dungeon,
  } })
  return data.data?.worldData?.encounter?.characterRankings?.rankings
}

const REPORT_SUMMARY = gql(/* GraphQL */`
  query ReportSummary($code: String!, $fight: Int!) {
    reportData {
      report(code: $code) {
        code
        table(fightIDs: [$fight], dataType: Summary)
      }
    }
}`)

interface StatsByPlayerReportsArgs {
  name: string
  server: string
  code: string
  fight: number
}

export interface CharacterStats {
  'strength'?: number
  'agility'?: number
  'intellect'?: number
  'stamina': number
  'leech': number
  'armor': number
  'mastery': number
  'haste': number
  'crit': number
  'versatility': number
  'item-level': number // note the quoted key for compatibility
  'avoidance': number
  'speed': number
  'block': number
  'parry': number
  'dodge': number
}

interface TopStats {
  name: string
  rank: number
  class: CharacterClass
  spec: Specialization
  stats: CharacterStats
}

interface StatRange {
  min: number
  max: number
}

interface PlayerStats {
  'Speed': StatRange
  'Mastery': StatRange
  'Crit': StatRange
  'Item Level': StatRange
  'Stamina': StatRange
  'Strength'?: StatRange
  'Agility'?: StatRange
  'Intellect'?: StatRange
  'Avoidance': StatRange
  'Haste': StatRange
  'Versatility': StatRange
  'Leech': StatRange
  'Armor': StatRange
}

// Type for CombatantInfo (Combatant-specific details)
interface CombatantInfo {
  stats: PlayerStats
}

// Type for PlayerSummary (General player information)
interface PlayerSummary {
  name: string
  id: number
  guid: number
  type: CharacterClass // Enum for character class
  server: string
  icon: string
  specs: Specialization[] // Array of specializations (using the Specialization enum)
  minItemLevel: number
  maxItemLevel: number
  potionUse: number
  healthstoneUse: number
  combatantInfo: CombatantInfo // Nested combatant info
}

export async function statsByPlayerReport({ name, server, code, fight }: StatsByPlayerReportsArgs): Promise<TopStats> {
  const client = await useClientWithEnv()
  const data = await client.query({ query: REPORT_SUMMARY, variables: { code, fight } })

  // find the player in report
  const players = data.data?.reportData?.report?.table.data.playerDetails as { tanks: PlayerSummary[], healers: PlayerSummary[], dps: PlayerSummary[] }
  server = server.replace(' ', '')
  const player = [players.tanks, players.dps, players.healers].flat().find(p => p.name === name && p.server === server)

  if (player === undefined) {
    throw new Error('Player not found in report')
  }

  // Extract keys and min values using JSONPath
  const keys = JSONPath<string[]>({ path: '$.*~', json: player.combatantInfo.stats }) // Extract keys
  const minValues = JSONPath<number[]>({ path: '$.*.min', json: player.combatantInfo.stats }) // Extract min values

  // Combine keys and values into a single object
  const stats = keys.reduce((acc, key, index) => {
    key = key.toLowerCase().replace(' ', '-')
    // @ts-expect-error: acc maybe of any type
    acc[key] = minValues[index]
    return acc
  }, {}) as CharacterStats

  return {
    name,
    class: player.type,
    spec: player.specs[0],
    rank: -1,
    stats,
  }
}

export async function topStatsBySpecDungeon({ className, specName, dungeon }: BySpecDungeonArgs): Promise<TopStats[]> {
  const details = await rankingsBySpecDungeon({ className, specName, dungeon })

  const top5Stats = details.slice(0, 5)

  // note that Promise.all preserve the order of its input array and output array
  return Promise.all(top5Stats.map(async (ts, i) => {
    const stats = await statsByPlayerReport({ name: ts.name, server: ts.server.name, code: ts.report.code, fight: ts.report.fightID })
    return {
      ...stats,
      rank: i,
    }
  }))
}
