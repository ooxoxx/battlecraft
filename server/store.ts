// models/TopStats.ts
import type { Optional } from 'sequelize'
import type { CharacterClass, CharacterStats, Dungeon, Specialization } from './types'
import { DataTypes, Model, Sequelize } from 'sequelize'

// Define the attributes for the TopStats model
interface TopStatsAttributes {
  id?: number
  name: string
  rank: number
  class: CharacterClass
  spec: Specialization
  dungeon: Dungeon
  stats: CharacterStats
}

// Define the input attributes for creation
interface TopStatsCreationAttributes extends Optional<TopStatsAttributes, 'id'> {}

// Define the Sequelize model
export class TopStatsStore extends Model<TopStatsAttributes, TopStatsCreationAttributes> implements TopStatsAttributes {
  public id!: number
  public name!: string
  public rank!: number
  public class!: CharacterClass
  public spec!: Specialization
  public dungeon!: Dungeon
  public stats!: CharacterStats

  // Timestamps
  public readonly createdAt!: Date
  public readonly updatedAt!: Date
}

// Initialize the model
function initTopStatsModel(sequelize: any) {
  TopStatsStore.init(
    {
      id: {
        type: DataTypes.INTEGER,
        autoIncrement: true,
        primaryKey: true,
      },
      name: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      rank: {
        type: DataTypes.INTEGER,
        allowNull: false,
      },
      class: {
        type: DataTypes.ENUM(
          'DeathKnight',
          'Druid',
          'Hunter',
          'Mage',
          'Monk',
          'Paladin',
          'Priest',
          'Rogue',
          'Shaman',
          'Warlock',
          'Warrior',
          'DemonHunter',
          'Evoker',
        ),
        allowNull: false,
      },
      spec: {
        type: DataTypes.ENUM(
          'Blood',
          'Frost',
          'Unholy',
          'Balance',
          'Feral',
          'Guardian',
          'Restoration',
          'BeastMastery',
          'Marksmanship',
          'Survival',
          'Arcane',
          'Fire',
          'Brewmaster',
          'Mistweaver',
          'Windwalker',
          'Holy',
          'Protection',
          'Retribution',
          'Discipline',
          'Shadow',
          'Assassination',
          'Subtlety',
          'Outlaw',
          'Elemental',
          'Enhancement',
          'Affliction',
          'Demonology',
          'Destruction',
          'Arms',
          'Fury',
          'Havoc',
          'Vengeance',
          'Devastation',
          'Preservation',
          'Augmentation',
        ),
        allowNull: false,
      },
      dungeon: {
        type: DataTypes.ENUM(
          '12660', // Ara-Kara, City of Echoes
          '12669', // City of Threads
          '60670', // Grim Batol
          '62290', // Mists of Tirna Scithe
          '61822', // Siege of Boralus
          '12662', // The Dawnbreaker
          '62286', // The Necrotic Wake
          '12652', // The Stonevault
        ),
        allowNull: false,
      },
      stats: {
        type: DataTypes.JSONB,
        allowNull: false,
      },
    },
    {
      sequelize,
      tableName: 'top_stats',
      indexes: [
        // Create a composite index for class, spec, and dungeon
        {
          name: 'class_spec_dungeon_index', // Name of the index
          fields: ['class', 'spec', 'dungeon'], // Fields to include in the index
          unique: false, // Set to true if the combination should be unique
        },
      ],
    },
  )
}

export const sequelize = new Sequelize({
  dialect: 'sqlite',
  storage: 'database.sqlite',
})

// Initialize models
initTopStatsModel(sequelize)

// Sync models with the database
sequelize.sync({ force: true }).then(() => {
  console.log('Database synced')
})

