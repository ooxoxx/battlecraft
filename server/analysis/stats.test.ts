import { expect, it } from 'vitest'
import { TopStatsStore } from '../store'
import { CharacterClass, Dungeon, Specialization } from '../types'
import { rankingsBySpecDungeon, statsByPlayerReport, topStatsBySpecDungeon } from './stats'

it('finds top k', async () => {
  const data = await rankingsBySpecDungeon({ className: CharacterClass.Hunter, specName: Specialization.BeastMastery, dungeon: Dungeon.AraKaraCityOfEchoes })
  // console.log(data.slice(0, 5))
  expect(data).toHaveLength(100)
})

it('finds stats by report', async () => {
  const data = await statsByPlayerReport({ name: '丶将息', server: '罗宁', code: 'GLQzYMwm1XDRV6Pn', fight: 54 })
  // console.log(data)
  expect(data).toHaveProperty('name')
  expect(data?.name).toEqual('丶将息')
  expect(data).toMatchSnapshot()
})

it('finds top players\' stats  by spec and dungeon.', async () => {
  const data = await topStatsBySpecDungeon({ className: CharacterClass.Hunter, specName: Specialization.BeastMastery, dungeon: Dungeon.AraKaraCityOfEchoes })
  expect(data).toHaveLength(5)
})

// it('aquires all top stats of all class/spec/dungeon combinations.', async () => {
//   await aquireTopStats()
//   const data = await TopStatsStore.findAll()
//   console.log(data.length)
// })
it ('stores all the top stats', async () => {
  const data = TopStatsStore.findAll()
  console.log(data)
})
