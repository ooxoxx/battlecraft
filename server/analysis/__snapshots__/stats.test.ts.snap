// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`finds stats by report 1`] = `
{
  "class": "Paladin",
  "dungeon": 12660,
  "name": "丶将息",
  "rank": -1,
  "spec": "Protection",
  "stats": {
    "armor": 122616,
    "avoidance": 0,
    "block": 0,
    "crit": 11903,
    "dodge": 0,
    "haste": 20800,
    "item-level": 639,
    "leech": 4577,
    "mastery": 12090,
    "parry": 11903,
    "speed": 684,
    "stamina": 516455,
    "strength": 58991,
    "versatility": 4085,
  },
}
`;
